docs/rfc/rfc-001-end-to-end-testing.md

```diff
@@ -0,0 +1,137 @@
+# RFC 001: End-to-End Testing
+
+## Changelog
+
+- 2020-09-07: Initial draft (@erikgrinaker)
+
+## Authors
+
+- Erik Grinaker (@erikgrinaker)
+
+## Context
+
+The current set of end-to-end tests under `test/` are very limited, mostly focusing on P2P testing in a standard configuration. They do not test various configurations (e.g. fast sync reactor versions, state sync, block pruning, genesis vs InitChain setup), nor do they test various network topologies (e.g. sentry node architecture). This leads to poor test coverage, which has caused several serious bugs to go unnoticed.
+
+We need an end-to-end test suite that can run a large number of combinations of configuration options, genesis settings, network topologies, ABCI interactions, and failure scenarios and check that the network is still functional. This RFC outlines the basic requirements and design considerations, but does not propose a specific implementation - a later ADR will be submitted for this.
+
+This RFC will not cover comprehensive chaos testing, only a few simple scenarios (e.g. abrupt process termination and network partitioning). Chaos testing of the core consensus algorithm should be implemented e.g. via Jepsen tests or a similar framework, or alternatively be added to these end-to-end tests at a later time. Similarly, malicious or adversarial behavior is out of scope for the first implementation, but may be added later.
+
+## Proposal
+
+### Functional Coverage
+
+The following lists the functionality we would like to test:
+
+#### Environments
+
+- Topology: single node, 4 nodes, sentry architecture, NAT (UPnP)
+
+- Networking: IPv4, IPv6
+
+- ABCI connection: UNIX socket, TCP, gRPC
+
+- PrivVal: file, UNIX socket, TCP
+
+#### Node/App Configurations
+
+- Database: goleveldb, cleveldb, boltdb, rocksdb, badgerdb
+
+- Fast sync: disabled, v0, v1, v2
+
+- State sync: disabled, enabled
+
+- Block pruning: none, keep 20, keep 1, keep random
+
+- Role: validator, full node
+
+### Geneses
```

Comment:
```suggestion
### Genesis
```

Reply:
Geneses is the plural form of genesis. Perhaps a bit too clever though. :)

Reply:
Haha realized that but didn't think it needed to be plural. You can ignore my comment.